keps/sig-network/2610-allports-services/README.md

```diff
@@ -0,0 +1,538 @@
+  - Estimated increase in size: (e.g., new annotation of size 32B)
+  - Estimated amount of new objects: (e.g., new Object X for every existing Pod)
+-->
+
+###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+
+<!--
+Look at the [existing SLIs/SLOs].
+
+Think about adding additional work or introducing new steps in between
+(e.g. need to do X to start a container), etc. Please describe the details.
+
+[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
+-->
+
+###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?
+
+<!--
+Things to keep in mind include: additional in-memory state, additional
+non-trivial computations, excessive access to disks (including increased log
+volume), significant amount of data sent and/or received over network, etc.
+This through this both in small and large cases, again with respect to the
+[supported limits].
+
+[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
+-->
+
+### Troubleshooting
+
+<!--
+This section must be completed when targeting beta to a release.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
```

Comment:
is an alternative to create a pure IP service, ie., it can also be a new Service Type IPOnly, RoutedIP (I'm bad at names :) and we just forward all IP packets to that IP, no matter port of protocol?
 can we rid of the ServicePorts field this way?

Reply:
I think this is the right way to implement an all-ip service. Otherwise at least SCTP must be included.

Reply:
I think this is the way to implement all-ip services but still with `type:LoadBalancer` but with no ports.

Reply:
That's a good idea. I wonder if this should be the main approach, rather than an alternative. The one big change is for LoadBalancer controllers to also process services of this new type. @thockin @rikatz any thoughts?

Reply:
A slight variation of the above could be:
```
apiVersion: v1
kind: Service
metadata:
  name: my-service
spec:
  selector:
    app: MyApp
  type: LoadBalancer
  protocols:
  - TCP
  - UDP
```

Reply:
- SCTP

Reply:
Explicitly declaring the protocols would help in cases where the application speaks protocols that are not TCP and UDP. 
```
protocols:
- SCTP
- L2TP
- 4in4
- 6in4
```

Reply:
This also makes it a future possibility to allow muti protocol services:
```
ports:
- protocol: TCP
  port: 6789
protocols:
- TCP
- ICMP
```
The default value of `protocols` can be populated from `ports`, (in most cases `[TCP]`)

Reply:
I added the "IPOnly" service type approach and "Loadbalancer + headless service + no ports" approach as alternatives. I think they both are achieving the same thing - Loadbalance at ip-level without specifying ports or protocol. But adding a new type of service is more overhead than modifying existing validation rules if they make sense.

I am not too familiar with the use-cases, but is opening up the loadbalancer service to all protocols problematic? I know iptables allows specifying -p "all" for all protocols and cloud-provider firewall rules should have a way to achieve the same as well. There is the security aspect of opening up all ports and protocols for a given LB IP with a very simple yaml change. 

We could do "LoadBalancer + headless service + no ports + new protocol list field" in that case. The downside is that LoadBalancer services are no longer guaranteed to be valid ClusterIP services.  LoadBalancer healthcheck does not depend on it..there could be other clients that expect a valid ClusterIP?

Reply:
> is an alternative to create a pure IP service, ie., it can also be a new Service Type IPOnly, RoutedIP
I like it :D

> The one big change is for LoadBalancer controllers to also process services of this new type.
I guess once this is supported in alpha, beta, etc cloud controllers should adopt this and maybe we can turn this as a graduation policy, like at least one cloud provider supporting the new type (although gatekeeping with cloud providers support might be much more harder to graduate the feature) 

About the new object type, If this service have the field ExternalIP it's still a LoadBalancer, but from a different type (as Antonio stated, a IPService). Anyway, I'm not sure all cloud providers support the LoadBalancer type without relying on NodePorts.

@thockin Maybe this could also define a subresource called ExternalIP, and with a subresource we can have different RBACs right? So would that be possible to say, for example 'only user X can set ExternalIP subresource in any service object'?

Reply:
resolving this, since this proposal is not creating a new service type anymore.